api/src/main/scala/za/co/absa/loginsvc/rest/service/AuthSearchService.scala

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2023 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.loginsvc.rest.service
+
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Service
+import za.co.absa.loginsvc.model.User
+import za.co.absa.loginsvc.rest.config.auth.{ActiveDirectoryLDAPConfig, UsersConfig}
+import za.co.absa.loginsvc.rest.config.provider.AuthConfigProvider
+import za.co.absa.loginsvc.rest.config.validation.ConfigValidationException
+
+import java.util
+import javax.naming.Context
+import javax.naming.directory.{Attributes, DirContext, SearchControls, SearchResult}
+import javax.naming.ldap.{Control, InitialLdapContext, PagedResultsControl}
+import scala.collection.JavaConverters.enumerationAsScalaIteratorConverter
+
+@Service
+class AuthSearchService @Autowired()(authConfigProvider: AuthConfigProvider) {
+
+  private val logger = LoggerFactory.getLogger(classOf[AuthSearchService])
+
+  private val usersConfig = authConfigProvider.getUsersConfig
+  private val adLDAPConfig = authConfigProvider.getLdapConfig
+
+  private val configs = Array(usersConfig, adLDAPConfig).filter(_.order != 0).sortBy(_.order)
+  if (configs.isEmpty)
+    throw ConfigValidationException("No authentication method enabled in config")
+
+  def searchUser(username: String): User = {
+
+    configs.foreach {
+      case u: UsersConfig =>
+        val result = searchUsersConfig(username)
+        if (result.isDefined) return result.get
+      case l: ActiveDirectoryLDAPConfig =>
+        val result = searchLDAP(username)
+        if (result.isDefined) return result.get
+      case _ => //ignore for now
+    }
+    throw new NoSuchElementException(s"Value not found in any object.")
+  }
```

Comment:
I understand that you are trying to short-circut here (not to lookup in LDAP if found in Config). If you redo the two search option into a common trait impls, there are ways to achieve that in a scala-friendly manner without `return`, e.g. using `takeWhile` - https://stackoverflow.com/a/13596556/1773349

Reply:
So Attempted to use a `takeWhile` `dropWhile` and `find` to accomplish this and while it works, it has the unfortunate side effect of having to run the a search function at least twice in the attempts I tried. So instead I separated the classes like you suggested with a common trait and just simplified this loop to something more readable.

Reply:
How about something like this (trying to do it in a Scala idiomatic fashion without using `return`):

```scala
  def searchUser(username: String): User = {
    val maybeFoundUser = orderedProviders.toIterator // this makes it lazy
      .map(_.searchForUser(username)) // expensive operation done lazily
      .dropWhile(_.isEmpty) // keeps searching until the possibly-found-user is not empty or all providers are exhausted

    if (maybeFoundUser.hasNext) {
      maybeFoundUser.next().getOrElse(throw new IllegalStateException(s"Valid user $maybeFoundUser should be available"))
    } else
      throw new NoSuchElementException(s"No user found by username $username anywhere")
  }
```

Related to this matter, I can see that `AuthSearchServiceTest` is present, but it only has a simple test for this case - `it should "return a matching user" in ...`. I think this class should be enriched to include the checks of multiple auth providers yielding different results:
 - ap1: Some, ap2: throw-should don't be called;
 -  ap1: None, ap2: Some;
 - ap1: None, ap2: None + observe the expected exception.

Reply:
I will actually try to do this, I have some other changes in mind - I will create a PR for this.

Reply:
I have created a PR that contains this https://github.com/AbsaOSS/login-service/pull/102

But it has other features - some classes broken down for testability and some example tests added to make the point. Please, give it a look

Reply:
Resolved by #102